File: doric-android/doric/src/main/java/pub/doric/plugin/NotchPlugin.java
package pub.doric.plugin;

import android.view.View;

import com.github.pengfeizhou.jscore.JSONBuilder;
import com.github.pengfeizhou.jscore.JSObject;
import com.github.pengfeizhou.jscore.JavaValue;
import com.qmuiteam.qmui.util.QMUINotchHelper;

import pub.doric.DoricContext;
import pub.doric.extension.bridge.DoricMethod;
import pub.doric.extension.bridge.DoricPlugin;
import pub.doric.extension.bridge.DoricPromise;
import pub.doric.utils.ThreadMode;

@DoricPlugin(name = "notch")
public class NotchPlugin extends DoricJavaPlugin {

    public NotchPlugin(DoricContext doricContext) {
        super(doricContext);
    }

    @DoricMethod(thread = ThreadMode.UI)
    public void inset(JSObject jsObject, final DoricPromise promise) {
        View view = getDoricContext().getRootNode().getNodeView();
        int top = QMUINotchHelper.getSafeInsetTop(view);
        int left = QMUINotchHelper.getSafeInsetLeft(view);
        int bottom = QMUINotchHelper.getSafeInsetBottom(view);
        int right = QMUINotchHelper.getSafeInsetRight(view);
        promise.resolve(new JavaValue(new JSONBuilder()
                .put("top", top)
                .put("left", left)
                .put("bottom", bottom)
                .put("right", right).toJSONObject()));
    }
}

File: doric-flutter/doric/lib/doric/plugin/notch_plugin.dart
import 'package:flutter/material.dart';
import 'package:doric/doric/engine/doric_plugin.dart';

import '../../flutter_jscore.dart';
import '../doric_context.dart';
import '../doric_promise.dart';

class NotchPlugin extends DoricPlugin {
  NotchPlugin(DoricContext context) : super(context);

  @override
  Map initFunc() {
    return {
      "inset": inset,
    };
  }

  @override
  void onTearDown() {}

  void inset(JSValue jsValue, DoricPromise promise) {
    var padding = MediaQuery.of(getDoricContext().getContext()).padding;
    promise.resolve([
      JSValue.makeFromObject(getDoricContext().getJSContext(), {
        "left": padding.left,
        "right": padding.right,
        "top": padding.top,
        "bottom": padding.bottom,
      })
    ]);
  }
}
